[store] shard not deleted after relocation if relocated shard is still in post recovery

brwe opened the pull request:
When deleting a shard the node that deletes the shard first checks if all shard copies are
started on other nodes. A message is sent to each node and each node checks locally for
STARTED or RELOCATED.
However, it might happen that the shard is still in state POST_RECOVERY, like this:

shard is relocating from node1 to node2
1. relocated shard on node2 goes in POST_RECOVERY and node2 sends shard started to master
2. master updates routing table and sends new cluster state to node1 and node2
3. node1 processes the cluster state and asks node2 if it has the active shard
  before node2 processes the new cluster state (which would cause it to set the shard to started)
4. node2 sends back it does not have the shard started and so node1 does not delete it

This can be avoided by waiting until cluster state that sets the shard to started is actually processed.

closes #10018

s1monw commented on src/main/java/org/elasticsearch/indices/store/IndicesStore.java:
```diff
@@ -385,6 +454,9 @@ public void writeTo(StreamOutput out) throws IOException {
             clusterName.writeTo(out);
             out.writeString(indexUUID);
             shardId.writeTo(out);
+            if (out.getVersion().onOrAfter(Version.V_1_5_0)) {
```
this needs to go away on master please

s1monw commented on src/main/java/org/elasticsearch/indices/store/IndicesStore.java:
```diff
@@ -333,8 +336,69 @@ public String executor() {
+            ShardId shardId = request.shardId;
+            IndexService indexService = indicesService.indexService(shardId.index().getName());
+            if (indexService != null && indexService.indexUUID().equals(request.indexUUID)) {
+                if (!indexService.hasShard(shardId.id())) {
```
can we use `== false` here?

s1monw commented on src/main/java/org/elasticsearch/indices/store/IndicesStore.java:
```diff
@@ -333,8 +336,69 @@ public String executor() {
+                channel.sendResponse(new ShardActiveResponse(false, clusterService.localNode()));
+            }
+            // create observer just in case
+            ClusterStateObserver observer = new ClusterStateObserver(clusterService, request.timeout, logger);
```
this observer needs to be closed I think unless you use it. I think you can just initialize it in the else clause?

brwe replied:
seems to me like there is no need to close and the close() method is unused anyway. I will remove it to avoid further confusion.

brwe replied:
see f253542abbf40c

s1monw commented on src/main/java/org/elasticsearch/indices/store/IndicesStore.java:
```diff
@@ -333,8 +336,69 @@ public String executor() {
+                        try {
+                            channel.sendResponse(new ShardActiveResponse(shardActive(request), clusterService.localNode()));
+                        } catch (IOException e) {
+                            logger.error("failed send response for shard active while trying to delete shard {} - shard will probably not be removed", request.shardId, new Exception());
```
the exception must be the first element after the log message then you can drop the second log line

s1monw commented on src/main/java/org/elasticsearch/indices/store/IndicesStore.java:
```diff
@@ -333,8 +336,69 @@ public String executor() {
+                        try {
+                            channel.sendResponse(new ShardActiveResponse(false, clusterService.localNode()));
+                        } catch (IOException e) {
+                            logger.error("failed send response for shard active while trying to delete shard {} - shard will probably not be removed", request.shardId, new Exception());
```
see above

s1monw commented on src/main/java/org/elasticsearch/indices/store/IndicesStore.java:
```diff
@@ -333,8 +336,69 @@ public String executor() {
+                        try {
+                            channel.sendResponse(new ShardActiveResponse(shardActive(request), clusterService.localNode()));
+                        } catch (IOException e) {
+                            logger.error("failed send response for shard active while trying to delete shard {} - shard will probably not be removed", request.shardId, new Exception());
```
see above

s1monw commented on src/main/java/org/elasticsearch/indices/store/IndicesStore.java:
```diff
@@ -333,8 +336,69 @@ public String executor() {
+            }
         }
 
         private boolean shardActive(ShardActiveRequest request) {
```
I think you have to pass the ClsuterState to this method otherwise processing is delayed. Just use the one from the predicate here.

brwe replied:
Not sure about that, I will check.

brwe replied:
Should not be a problem. `IndicesClusterStateService` is in the `ClusterService.priorityClusterStateListeners` and the observer is in the `ClusterService.postAppliedListeners` which are notified afterwards so we can be sure the shard has moved to started. In addition, the ClusterState does not tell if the shard is actually moved to started.

s1monw commented:
cool stuff I left some comments

brwe commented:
thanks for the review! addressed all comments. want to take another look?

s1monw commented on src/main/java/org/elasticsearch/indices/store/IndicesStore.java:
```diff
@@ -333,8 +336,66 @@ public String executor() {
+                        try {
+                            channel.sendResponse(new ShardActiveResponse(shardActive(request), clusterService.localNode()));
+                        } catch (IOException e) {
+                            logger.error("failed send response for shard active while trying to delete shard {} - shard will probably not be removed", e, request.shardId, new Exception());
```
this log line has still an obsolete new Exception?

s1monw commented on src/main/java/org/elasticsearch/indices/store/IndicesStore.java:
```diff
@@ -333,8 +336,66 @@ public String executor() {
+                        try {
+                            channel.sendResponse(new ShardActiveResponse(false, clusterService.localNode()));
+                        } catch (IOException e) {
+                            logger.error("failed send response for shard active while trying to delete shard {} - shard will probably not be removed", e, request.shardId, new Exception());
```
this log line has still an obsolete new Exception?

s1monw commented on src/main/java/org/elasticsearch/indices/store/IndicesStore.java:
```diff
@@ -333,8 +336,66 @@ public String executor() {
+                        try {
+                            channel.sendResponse(new ShardActiveResponse(shardActive(request), clusterService.localNode()));
+                        } catch (IOException e) {
+                            logger.error("failed send response for shard active while trying to delete shard {} - shard will probably not be removed", e, request.shardId, new Exception());
```
this log line has still an obsolete new Exception?

s1monw commented:
left some 3 comments other than that LGTM

brwe commented:
Pushed another commit. We have to catch EsRejectedExecutionException when we try to send back whether the shard is active or not. For example, InternalNode.stop will cause  ObserverClusterStateListener.onClose of the listener to be called at some point and the reject exception that this might throw is not caught anywhere it seems. Alternatively we might also consider not trying to send back a response on close?

s1monw commented:
I think we should catch there exception in the caller of the close method - this can hit the next user of this as well?

s1monw commented on src/main/java/org/elasticsearch/indices/store/IndicesStore.java:
```diff
@@ -333,8 +337,63 @@ public String executor() {
+                        sendResultOrLogError(shardActive(request));
+                    }
+
+                    public void sendResultOrLogError(boolean shardActive) {
```
can this method just be called `sendResponse` or so?

bleskes commented on src/main/java/org/elasticsearch/indices/store/IndicesStore.java:
```diff
@@ -333,8 +337,63 @@ public String executor() {
+                            logger.error("failed send response for shard active while trying to delete shard {} - shard will probably not be removed", e, request.shardId);
+                        }
+                    }
+                }, new ClusterStateObserver.ChangePredicate() {
```
can we inherit ValidationPredicate ? it takes care of all the subtleties and requires implementing one method

bleskes commented on src/main/java/org/elasticsearch/indices/store/IndicesStore.java:
```diff
@@ -333,8 +337,63 @@ public String executor() {
+        public void messageReceived(final ShardActiveRequest request, final TransportChannel channel) throws Exception {
+            // make sure shard is really there before register cluster state obsever
+            ShardId shardId = request.shardId;
+            IndexService indexService = indicesService.indexService(shardId.index().getName());
```
this looks very much like the code in shardActive

brwe replied:
There is two tasks: get the shard and check if it is active. I changed that now to have two explicit methods for this to avoid code duplication, see 01e40f8d024e7965ab8914755f183a269908eaaa But I do not understand how the AsyncShardOperationAction#retry pattern would help here because I do not think we actually want to recurse?

bleskes commented on src/test/java/org/elasticsearch/indices/store/IndicesStoreIntegrationTests.java:
```diff
@@ -95,6 +95,12 @@ public void indexCleanup() throws Exception {
         logger.info("--> move shard from node_1 to node_3, and wait for relocation to finish");
+        SlowClusterStateProcessing disruption = null;
+        if (randomBoolean()) {
+            disruption = new SlowClusterStateProcessing(node_3, getRandom(), 0, 0, 1000, 2000);
```
do we need random sleeps here or can we use BlockClusterStateProcessing to make this more reproducible?

brwe replied:
I am unsure how. I want to check if it will be deleted even after a small delay, so I would have to put a sleep between startDisruption() and stopDisruption() and this around the reroute below and that does not seem right. Or did you have something else in mind?

bleskes replied:
Oh i see now what you do. You want to make sure the shard is deleted even if CS processing is slow (or but not missing). I think its ok now. It will be hard to make this work with an absolute block as it will also be in the way of the relocation command.

bleskes commented on src/main/java/org/elasticsearch/indices/store/IndicesStore.java:
```diff
@@ -357,7 +416,8 @@ private boolean shardActive(ShardActiveRequest request) {
 
     private static class ShardActiveRequest extends TransportRequest {
-
+        public static final TimeValue DEFAULT_TIMEOUT = new TimeValue(1, TimeUnit.SECONDS);
```
I think we want to make this less aggressive, no? how about 30s? also, I wonder if we should make it configurable to be safe and allow us to speed it up in tests (imho it doesn't have to be dynamic) .

bleskes commented:
I like the approach. Left some comments.

brwe commented:
@bleskes @s1monw thanks a lot for the review! I implemented all changes except where I added a comment because I was unsure what to do. 
@s1monw about the exception handling: It seems in general unchecked exceptions are not handled when listeners are called when the cluster service closes. I can add a catch for them (see b4e88ed9a155a9cd3832403b574efbdf9db612eb) but because they are not handled anywhere I suspect there is method behind it. I'd be happy for any insight into how exceptions should be handled properly here.

s1monw commented:
@brwe I think we should detach the exception handling problem from this issue. Yet, we should still address it. IMO we really need to make sure that all listeners are notified even if one of them threw an exception. Can you open a folllowup?

s1monw commented:
other than that ^^ LGTM :)

bleskes commented on src/test/java/org/elasticsearch/indices/store/IndicesStoreIntegrationTests.java:
```diff
@@ -108,6 +114,9 @@ public void indexCleanup() throws Exception {
         assertThat(Files.exists(indexDirectory(node_2, "test")), equalTo(true));
         assertThat(Files.exists(shardDirectory(node_3, "test", 0)), equalTo(true));
         assertThat(Files.exists(indexDirectory(node_3, "test")), equalTo(true));
+        if (disruption != null) {
```
not needed. The test infra cleans up disruptions..

bleskes commented on src/test/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -82,6 +82,7 @@
 import org.elasticsearch.indices.breaker.HierarchyCircuitBreakerService;
 import org.elasticsearch.indices.fielddata.cache.IndicesFieldDataCache;
 import org.elasticsearch.indices.recovery.RecoverySettings;
+import org.elasticsearch.indices.store.IndicesStore;
```
extra import

bleskes commented on src/main/java/org/elasticsearch/indices/store/IndicesStore.java:
```diff
@@ -333,31 +341,82 @@ public String executor() {
+            if (indexShard == null) {
+                channel.sendResponse(new ShardActiveResponse(false, clusterService.localNode()));
+            }
+            // create observer just in case
```
we don't use this now here, right? no need to create just in case.

bleskes commented:
LGTM. Left some very minor comments. no need for another review cycle.

brwe commented:
pushed to master and 1.x (17dffe222b923c17614905515773614d6963e13e)
